blog/tests/test_blog_article_page.py

```diff
@@ -71,3 +68,12 @@ def test_that_title_should_not_be_longer_than_custom_specified_amount_of_charact
         title_that_is_way_too_long = "f" * (MAX_BLOG_ARTICLE_TITLE_LENGTH + 1)
         with self.assertRaises(ValidationError):
             self._create_blog_article_page(title=title_that_is_way_too_long)
+
+    def test_that_reccomended_articles_should_not_contain_duplicates(self):
+        other_article = self._create_blog_article_page()
+        articles_block = StreamBlock([("page", PageChooserBlock())])
+        recommended_articles_with_duplicates = StreamValue(
+            articles_block, [("page", other_article), ("page", other_article)]
+        )
+        with self.assertRaises(ValidationError):
+            self._create_blog_article_page(recommended_articles=recommended_articles_with_duplicates)
```

Comment:
There is still lack of test which explicitly tests that there is no recommended article added to new Article page and also please add test for case when you try to add duplicated recommended article when this article already exists and is saved and stored in database. Such things must be tested explicitly.

Reply:
Aren't those two cases already covered though?

Recommendations are omitted in the **test_that_new_article_page_should_has_all_mandatory_parameters** defaulting to None in the **_create_blog_article_page** helper function.
The description of the test is key here - **mandatory_parameters** implies that the **recommended_articles** is optional. The test passes <=> having no recommended articles is valid.
There really isn't any custom logic here, it's all Wagtail's default documented behavior. If I were to write an explicit test for the case of no recommendations provided, we might as well write similar tests for other optional fields, like photographs.

The duplicate test goes as follows:
```
        other_article = self._create_blog_article_page()
        articles_block = StreamBlock([("page", PageChooserBlock())])
        recommended_articles_with_duplicates = StreamValue(
            articles_block, [("page", other_article), ("page", other_article)]
        )
        with self.assertRaises(ValidationError):
            self._create_blog_article_page(recommended_articles=recommended_articles_with_duplicates)
```
>test for case when you try to add duplicated recommended article when this article already exists and is saved and stored in database

**_create_blog_article_page** creates a model and performs **save()** on it, therefore it is normally stored in the database like it normally should be, right?
So when the rest of the test is performed that duplicated article **already exists** in the database, therefore your criteria is fulfilled?

Could you please elaborate what is missing in the existing approach? Because I am honestly confused as I fail to see the error. Have I maybe misunderstood your comment?

Reply:
Done.